pyfar/dsp/dsp.py

```diff
@@ -227,6 +227,219 @@ def spectrogram(signal, dB=True, log_prefix=20, log_reference=1,
+        elif unit == 'ms':
+            length = [round(li*signal.sampling_rate/1e3) for li in length]
+        elif unit == 'samples':
+            length = [int(li) for li in length]
+        else:
+            raise ValueError(f"unit is {unit} but has to be"
+                             f" 'samples', 's' or 'ms'.")
+        # Check window size
+        if length[-1] > signal.n_samples:
+            raise ValueError(
+                "Values in length require window to be longer than signal.")
+
+    # Create window
+    # win_start and win_stop define the first and last sample of the window
+    if length is None:
+        win_samples = signal.n_samples
+        win = sgn.windows.get_window(window, win_samples, fftbins=False)
+        win_start = 0
+        win_stop = signal.n_samples-1
+    elif len(length) == 2:
+        if shape == 'symmetric':
+            win_samples = length[1]-length[0]+1
+            win = sgn.windows.get_window(window, win_samples, fftbins=False)
+            win_start = length[0]
+            win_stop = length[1]
+        else:
+            fade_samples = int(2*(length[1]-length[0]))
+            fade = sgn.windows.get_window(window, fade_samples, fftbins=False)
+            if shape == 'symmetric_zero':
+                win = np.zeros(signal.n_samples)
+                win[:length[0]+1] = 1
+                win[length[0]+1:length[1]+1] = fade[int(fade_samples/2):]
+                win[-length[0]:] = 1
+                win[-length[1]:-length[0]] = fade[:int(fade_samples/2)]
+                win_start = 0
+                win_stop = signal.n_samples
+            elif shape == 'left':
+                win = np.ones(signal.n_samples-length[0])
+                win[0:length[1]-length[0]] = fade[:int(fade_samples/2)]
+                win_start = length[0]
+                win_stop = signal.n_samples-1
+            elif shape == 'right':
+                win = np.ones(length[1]+1)
+                win[length[0]+1:] = fade[int(fade_samples/2):]
+                win_start = 0
+                win_stop = length[1]
+    elif len(length) == 4:
+        fade_in_samples = int(2*(length[1]-length[0]))
+        fade_in = sgn.windows.get_window(
+            window, fade_in_samples, fftbins=False)
+        fade_in = fade_in[:int(fade_in_samples/2)]
+        fade_out_samples = int(2*(length[3]-length[2]))
+        fade_out = sgn.windows.get_window(
+            window, fade_out_samples, fftbins=False)
+        fade_out = fade_out[int(fade_out_samples/2):]
+        win = np.ones(length[-1]-length[0]+1)
+        win[0:length[1]-length[0]] = fade_in
+        win[length[2]-length[0]+1:length[3]-length[0]+1] = fade_out
+        win_start = length[0]
+        win_stop = length[3]
+    else:
+        raise ValueError(
+            "Length needs to contain two or four values.")
+
+    # Apply window
+    signal_win = signal.copy()
+    if crop:
+        signal_win.time = signal_win.time[..., win_start:win_stop+1]*win
```

Comment:
One more thing I just realized: If a crop is applied in the beginning of the signal, we also apply a linear-phase part with a negative delay of `np.exp(-1j*win_start)`. This may be potentially unwanted (for example for HRTF measurments, or any other measurments where the original phase needs to be preserved). I would suggest to have more options for the crop operation, for example `'full'` which is the current version, `'none'` which doesn't crop at all, and an additional option which crops to the end of the fade out window. I don't have a good one word description though ;)

Reply:
Good point! Added three options ``'none'``, ``'window'`` and ``'end'``. Any ideas for better descriptions?

Reply:
I like the three you chose